[TextareaAutosize] Show scrollbar when rowsMax is exceeded

Shubhamchinda opened the pull request:
Fixes #17289 
@oliviertassinari

oliviertassinari commented:
@Shubhamchinda Could you add a test case that demonstrate that the fix is working?

Shubhamchinda commented:
> @Shubhamchinda Could you add a test case that demonstrate that the fix is working?

Yes for sure. @oliviertassinari 
I need to make these changes?
```diff
it('should update when uncontrolled', () => {
      const handleChange = spy();
      const wrapper = mount(<TextareaAutosize onChange={handleChange} />);
-     assert.deepEqual(getStyle(wrapper), { height: undefined, overflow: null });
+     assert.deepEqual(getStyle(wrapper), { height: 0, overflow: "hidden" });
```
in `packages/material-ui/src/TextareaAutosize/TextareaAutosize.js b/packages/material-ui/src/TextareaAutosize/TextareaAutosize.test.js`

Let me know if this is correct!

Thanks!

oliviertassinari commented:
@Shubhamchinda You need to
1. handle the existing tests, make sure the new logic is correct, and if it's the case update the test according
2. add a new test that would fail without the fix and pass with it, something that help avoid a regression in the future.

Shubhamchinda commented:
> @Shubhamchinda You would need to
> 
> 1. handle the existing tests, make sure the new logic is correct, and if it's the case update the test according.
> 2. add a new test that would fail without the fix and pass with it, something that helps avoid regressions in the future.

Okay, Thanks a lot. :)
I will update the PR.

Shubhamchinda commented:
Hey @oliviertassinari ,always good to hear from you.

I updated the test cases. Please take a look and let me know if I did it correctly.

Shubhamchinda commented on packages/material-ui/src/TextareaAutosize/TextareaAutosize.test.js:
```diff
@@ -79,8 +79,8 @@ describe('<TextareaAutosize />', () => {
-          height: undefined,
-          overflow: null,
+          height: 0,
+          overflow: 'hidden',
```
I don't think there should be any changes here, So I think maybe We need to come up another logic.
I guess `overflow` should be `null` here.

What are your thoughts?

Shubhamchinda commented on packages/material-ui/src/TextareaAutosize/TextareaAutosize.test.js:
```diff
@@ -102,7 +102,7 @@ describe('<TextareaAutosize />', () => {
       const handleChange = spy();
       const wrapper = mount(<TextareaAutosize onChange={handleChange} />);
-      assert.deepEqual(getStyle(wrapper), { height: undefined, overflow: null });
+      assert.deepEqual(getStyle(wrapper), { height: 0, overflow: 'hidden' });
```
Same here

Shubhamchinda commented on packages/material-ui/src/TextareaAutosize/TextareaAutosize.test.js:
```diff
@@ -122,7 +122,7 @@ describe('<TextareaAutosize />', () => {
       const border = 5;
       const wrapper = mount(<TextareaAutosize />);
-      assert.deepEqual(getStyle(wrapper), { height: undefined, overflow: null });
+      assert.deepEqual(getStyle(wrapper), { height: 0, overflow: 'hidden' });
```
Same here.

Shubhamchinda commented on packages/material-ui/src/TextareaAutosize/TextareaAutosize.test.js:
```diff
@@ -184,6 +184,22 @@ describe('<TextareaAutosize />', () => {
+      });
+      wrapper.setProps();
+      wrapper.update();
+      assert.deepEqual(getStyle(wrapper), { height: lineHeight * rowsMax, overflow: null });
```
`overflow` should be `hidden` here, right?

Shubhamchinda commented on packages/material-ui/src/TextareaAutosize/TextareaAutosize.js:
```diff
@@ -68,17 +68,18 @@ const TextareaAutosize = React.forwardRef(function TextareaAutosize(props, ref)
         return {
-          innerHeight,
-          outerHeight,
+          overflow,
```
Sorry for so many comments.

We're returing `overflow` even if the below condition is true.
```
 (outerHeightStyle > 0 &&
        Math.abs((prevState.outerHeightStyle || 0) - outerHeightStyle) > 10)
```

Don't you think `overflow` state should only be updated when `prevState.overflow !== overflow` is true ?

oliviertassinari replied:
We are using a single state. I wouldn't expect any difference.

Shubhamchinda replied:
Umm Okay!

Shubhamchinda commented:
> * I have rebased on the master branch to make the Netlify deploy pass
> * I have updated the test to fail without the fix
> * The changes in the test files look harmless, I'm not aware of any regression

Thanks a lot @oliviertassinari :)
Though I have a doubt in  https://github.com/mui-org/material-ui/pull/17310#discussion_r321600930 could you please clear it? I'd really appreciate it.

Thanks again :)

oliviertassinari commented:
Thanks
